estela_scrapy/extensions.py

```diff
@@ -65,19 +53,64 @@ def item_scraped(self, item):
         }
         producer.send("job_items", data)
 
+
+class RedisStatsCollector(BaseExtension):
+    def __init__(self, stats):
+        super().__init__(stats)
+
+        redis_url = os.getenv("REDIS_URL")
+        if not redis_url:
+            raise NotConfigured("REDIS_URL not found in the settings")
+        self.redis_conn = redis.from_url(redis_url)
+
+        self.stats_key = os.getenv("REDIS_STATS_KEY")
+        self.interval = float(os.getenv("REDIS_STATS_INTERVAL"))
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        ext = cls(crawler.stats)
+
+        crawler.signals.connect(ext.spider_opened, signal=signals.spider_opened)
+        crawler.signals.connect(ext.spider_closed, signal=signals.spider_closed)
+
+        return ext
+
+    def spider_opened(self, spider):
+        self.task = task.LoopingCall(self.store_stats, spider)
+        self.task.start(self.interval)
+
     def spider_closed(self, spider, reason):
-        spider_stats = self.stats.get_stats()
-        self.update_job(
+        if self.task.running:
+            self.task.stop()
+
+        try:
+            self.redis_conn.delete(self.stats_key)
+        except Exception:
+            pass
+
+        stats = self.stats.get_stats()
+        update_job(
+            self.job_url,
+            self.auth_token,
             status=COMPLETED_STATUS,
-            lifespan=spider_stats.get("elapsed_time_seconds", 0),
-            total_bytes=spider_stats.get("downloader/response_bytes", 0),
-            item_count=spider_stats.get("item_scraped_count", 0),
-            request_count=spider_stats.get("downloader/request_count", 0),
+            lifespan=int(stats.get("elapsed_time_seconds", 0)),
+            total_bytes=stats.get("downloader/response_bytes", 0),
+            item_count=stats.get("item_scraped_count", 0),
+            request_count=stats.get("downloader/request_count", 0),
         )
 
-        parser_stats = json.dumps(spider_stats, default=json_serializer)
+        parsed_stats = json.dumps(stats, default=json_serializer)
         data = {
             "jid": os.getenv("ESTELA_SPIDER_JOB"),
-            "payload": json.loads(parser_stats),
+            "payload": json.loads(parsed_stats),
         }
-        producer.send("job_stats", data)
+        producer.send("job_stats", value=data).add_errback(on_kafka_send_error)
+        producer.flush()
```

Comment:
Where is the variable `on_kafka_send_error`?

Reply:
I missed that. Nice catch.